File: Scripts/Global/Network/PacketUtil.cs
using System;
using System.Runtime.InteropServices;

namespace NetworkPackets
{
    enum PACKET_TYPE : byte
    {
        ///////////////////////////////Client와의 통신
        REQUEST_LOGIN = 0,     //login request 
        RESPONSE_LOGIN = 1,     //login connect 
        DISCONNECTED_LOGIN_SYSTEM = 2,     //logout  
        HELLO_UDP = 3,     //helloUdp
        HELLO_UDP_RESPONSE = 4,     //helloUdp response

        ///////////////////////////////UpperServer와 통신
        REQUEST_CONNECT = 5,     //upperserver와 연결요청
        RESPONSE_CONNECT = 6,     //upperserver의 응답
        UPDATE_ROUTING_INFO = 7,  //upperserver에게 routing path info 전달

        ///////////////////////////////모의통제도구 와 통신
        REQUEST_SIM_START = 10,    //시뮬레이션 시작 요청
        REQUEST_SIM_STOP = 11,    //시뮬레이션 중단 요청


        RESPONSE_SIM_START = 12,    //시뮬레이션 시작 응답(broadcast)


        ///////////////////////////////전장상황도 와 통신
        REQUEST_MILITARY_UNIT_INFO = 20,   //부대정보 요청
        RESPONSE_MILITARY_UNIT_INFO = 21,   //부대정보 응답

        REQUEST_ORDER_IN_SALLY = 31,   //출격명령 요청
        RESPONSE_ORDER_IN_SALLY = 32,   //출격명령 응답
        REQUEST_ORDER_IN_ATTACK = 33,   //발사명령 요청
        RESPONSE_ORDER_IN_ATTACK = 34,   //발사명령 응답


        SYNC_ARTILLERY_COORDINATE = 45,   //자주포객체의 좌표 동기화(udp 통신)
        SYNC_MISSILE_COORDINATE = 46,   //미사일객체의 좌표 동기화(udp 통신)
        SYNC_ARTILLERY_ATTACKED = 47,
        SYNC_MILITARY_UNIT_ATTACKED = 48,

        ///////////////////////////////시뮬레이션 시간 키핑관련 통신
        SYNC_SIM_TIME = 50,   //시뮬레이션 엔진과 시간 동기화(broadcast)
    };

    enum USER_AUTHORITY : byte
    {
        USER = 0,     //user 
        ADMIN = 1,     //admin 
    };
    enum MILITARY_ID : ushort
    {
        ADMIN_ID  = 0,     //모든 부대아이디 .
        DEFAULT   = 1,     //default 부대 id
    };
    enum NETWORK_FLAG : byte
    {
        UDP = 0,     //udp
        TCP = 1,     //tcp
        BROADCAST_UDP = 2,
        BROADCAST_TCP = 3,


        ///////////////////////////load obj 관련 flag
        NONE          = 10,
        FIN           = 11,

        ///////////////////////////login 관련 flag
        LOGIN_SUCCESS     = 20,
        LOGIN_NO_FOUND_ID = 21,
        LOGIN_NO_FOUND_PW = 22,

        SERVER_ERROR      = 30,
    };



    // 공통 헤더
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StPacketHeader
    {
        public byte nType;          // uint8_t
        public int nID;             // int32_t
        public byte nFlagField;     // uint8_t
    }

    // 군부대 정보
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StMilitaryPacketInfo
    {
        public ushort nID;          // uint16_t
        public uint nTotalArtillery;// uint32_t
        public uint nXpos;          // uint32_t
        public uint nYpos;          // uint32_t
        public uint nUnitHealth;    // uint32_t
    }

    // 자주포 정보
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StArtilleryPacketInfo
    {
        public ushort nID;          // uint16_t
        public uint nTotalMissile;   // uint32_t
        public uint nXpos;          // uint32_t
        public uint nYpos;          // uint32_t
        public uint nUnitHealth;    // uint32_t
    }

    // 미사일 정보
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StMissilePacketInfo
    {
        public uint nID;            // uint32_t
        public uint nXpos;          // uint32_t
        public uint nYpos;          // uint32_t
    }


    // 로그인 요청
    [StructLayout(LayoutKind.Sequential, Pack = 1, CharSet = CharSet.Ansi)]
    public struct StLoginRequestPacketData
    {
        public StPacketHeader strtHeader;

        [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 64)]
        public string nUserId;

        [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 64)]
        public string nUserPw;
    }

    // 로그인 응답
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StLoginResponsePacketData
    {
        public StPacketHeader strtHeader;
        public int nUserClientFD;
        public sbyte nUserAuthority;        // int8_t
        public short nUserMilitarySection;  // int16_t
    }

    // 엔진 시작 요청
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StStartRequestEnginePacket
    {
        public StPacketHeader strtHeader;
    }

    // 엔진 시작 응답
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StStartResponseEnginePacket
    {
        public StPacketHeader strtHeader;
    }

    // 군부대 정보 요청
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StRequestLoadMilitaryUnitInfoPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;          // int32_t
        public short nUserMilitarySection; // int16_t
    }

    // 군부대 정보 응답
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StResponseLoadMilitaryUnitInfoPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;                // int32_t
        public StMilitaryPacketInfo strtMilitaryInfo;
        public byte nIsFin;
    }

    // 출격 요청
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StRequestOrderInSallyPacket
    {
        public StPacketHeader strtHeader;
        public short nUserMilitarySection;   // int16_t
        public int nUserClientFd;            // int32_t
        public int nDstXPos;                 // int32_t
        public int nDstYPos;                 // int32_t
    }

    // 출격 응답
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StResponseOrderInSallyPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;                  // int32_t
        public StMilitaryPacketInfo strtMilitaryInfo;
        public StArtilleryPacketInfo strtArtilleryInfo;
    }

    // 미사일 발사 요청
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StRequestOrderInAttackPacket
    {
        public StPacketHeader strtHeader;
        public short nArtilleryId;           // int16_t
        public int nUserClientFd;            // int32_t
        public int nDstXPos;                 // int32_t
        public int nDstYPos;                 // int32_t
    }

    // 미사일 발사 응답
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StResponseOrderInAttackPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;                  // int32_t
        public StArtilleryPacketInfo strtArtilleryInfo;
        public StMissilePacketInfo strtMissileInfo;
    }


    // 자주포 좌표 동기화
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StSyncArtilleryCoordinatePacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;            // int32_t
        public StArtilleryPacketInfo strtArtilleryInfo;
    }

    // 미사일 좌표 동기화
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StSyncMissileCoordinatePacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;            // int32_t
        public StMissilePacketInfo strtMissileInfo;
    }

    // 자주포   attacked 동기화
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StSyncArtilleryAttackedPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;            // int32_t
        public StArtilleryPacketInfo strtArtilleryInfo;
    }

    // military unit attacked 동기화
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StSyncMilitaryUnitAttackedPacket
    {
        public StPacketHeader strtHeader;
        public int nUserClientFd;            // int32_t
        public StMilitaryPacketInfo strtMilitaryInfo;
    }

    // 시뮬레이션 시간 동기화
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StSyncSimulationTimePacket
    {
        public StPacketHeader strtHeader;
        public int nCurrentSimulationTime;   // int32_t
    }

    // 기본 패킷 (헤더만)
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct StPacket
    {
        public StPacketHeader strtHeader;
    }
}

File: NetworkModule/NetworkModule/PacketUtil.cpp
// ********************************************************************
// * 소스정의: PacketUtil.h
// * 설    명: packet protocol 과 관련한 유틸함수들을 구현한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 10
// ********************************************************************
#include "pch.h"
#include "PacketUtil.h"


using namespace DATATRSFER;

// ********************************************************************
// * 함 수 명: ParsePacketInfo
// * 설    명: 패킷을 파싱한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 10
// ********************************************************************
int_t PacketUtil::ParsePacketInfo(const char_t* pPacket)
{
    return -1;
}

// ********************************************************************
// * 함 수 명: CreateResponseLoginpacket
// * 설    명: accept후에 클라이언트가 보낸 로그인 요청응답
// *           1.   로그인 성공 실패 여부 반환
// *           2.   성공시 clientFd(clientID)전송후 항상 포함하도록...
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 10
// ********************************************************************
void PacketUtil::CreateResponseLoginPacket(int32_t nClientFd, struct _Packet* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::RESPONSE_LOGIN;
    pStrtPacket->strtHeader.nID = htonl(nClientFd);
}

// ********************************************************************
// * 함 수 명: CreateRequestLoginPacket
// * 설    명: accept후에 user에게 로그인 정보를 요청
// *           clientFd(clientID)전송후 항상 포함하도록...
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 10
// ********************************************************************
void PacketUtil::CreateRequestLoginPacket(const char_t* strUserId,const char_t* strUserPW,StLoginRequestPacketData* pStrtPacket)
{
    // 전체 초기화
    memset(pStrtPacket, 0, sizeof(StLoginRequestPacketData));


    pStrtPacket->strtHeader.nType = PACKET_TYPE::REQUEST_LOGIN;
    pStrtPacket->strtHeader.nID = htonl(CLIENT_TYPE::INTEGRATE_LOGIN_SYSTEM);
    
    // User ID 채우기 (최대 63바이트 + 널 문자)
    strncpy_s(pStrtPacket->nUserId, strUserId, sizeof(pStrtPacket->nUserId) - 1);
    pStrtPacket->nUserId[sizeof(pStrtPacket->nUserId) - 1] = '\0';

    // User PW 채우기 (최대 63바이트 + 널 문자)
    strncpy_s(pStrtPacket->nUserPw, strUserPW, sizeof(pStrtPacket->nUserPw) - 1);
    pStrtPacket->nUserPw[sizeof(pStrtPacket->nUserPw) - 1] = '\0';

}

// ********************************************************************
// * 함 수 명: CreateRequettHelloUdp
// * 설    명: 로그인 성공 후에 client가 보낼 HelloUdp
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 11
// ********************************************************************
void PacketUtil::CreateRequestHelloUdp(int32_t nClientFd, struct _Packet* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::HELLO_UDP;
    pStrtPacket->strtHeader.nID = htonl(nClientFd);

}

// ********************************************************************
// * 함 수 명: CreateResponseHelloUdp
// * 설    명: 로그인 성공 후에 client가 보낸 helloUDP에 대한 응답
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 10
// ********************************************************************
void PacketUtil::CreateResponseHelloUdp(int32_t nClientFd, struct _Packet* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::HELLO_UDP_RESPONSE;
    pStrtPacket->strtHeader.nID = htonl(nClientFd);

}

// ********************************************************************
// * 함 수 명: CreateRequestSimulationStartPacket
// * 설    명: SimEngine과 연결된 상위 데이터전달처리기에 시뮬레이션 시작 
// *           패킷을 보낸다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 30
// ********************************************************************
void PacketUtil::CreateRequestSimulationStartPacket(int32_t nClientFd, StStartRequestEnginePacket* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::REQUEST_SIM_START;
    pStrtPacket->strtHeader.nID = htonl(nClientFd);

}

// ********************************************************************
// * 함 수 명: CreateResponseSimulationStartPacket
// * 설    명: broadcast 로 모든 클라이언트에게 시뮬레이션 시작을 알린다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 30
// ********************************************************************
void PacketUtil::CreateResponseSimulationStartPacket(StStartResponseEnginePacket* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::RESPONSE_SIM_START;
    //pStrtPacket->strtHeader.nID   = htonl(nClientFd);

}

// ********************************************************************
// * 함 수 명: CreateRequestLoadMilitaryUnitPacket
// * 설    명: 부대id에 맞게 필요한 객체정보를 요청한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 30
// ********************************************************************
void PacketUtil::CreateRequestLoadMilitaryUnitPacket(int32_t nThisFd, int32_t nClientFd, int16_t nMilitaryId, StRequestLoadMilitaryUnitInfoPacket* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::REQUEST_MILITARY_UNIT_INFO;
    pStrtPacket->strtHeader.nID = htonl(nThisFd);
    pStrtPacket->nUserClientFd = htonl(nClientFd);
    pStrtPacket->nUserMilitarySection = htons(nMilitaryId);
    DEBUG("원본 militaryId : " + std::to_string(nMilitaryId));
    DEBUG("send militaryId : " + std::to_string(pStrtPacket->nUserMilitarySection));
}

// ********************************************************************
// * 함 수 명: CreateResponseLoadMilitaryUnitPacket
// * 설    명: 부대id에 맞게 필요한 객체정보를 요청한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 07. 31
// ********************************************************************
void PacketUtil::CreateResponseLoadMilitaryUnitPacket(int32_t nClientFd, const StResponseLoadMilitaryUnitInfoPacket* pRecvPacket, StResponseLoadMilitaryUnitInfoPacket* pSendPacket)
{
    //*pSendPacket = *pRecvPacket;

    pSendPacket->strtHeader.nID = htonl(nClientFd);
    pSendPacket->strtHeader.nType = PACKET_TYPE::RESPONSE_MILITARY_UNIT_INFO;
    pSendPacket->nUserClientFd = pRecvPacket->nUserClientFd;
    pSendPacket->strtMilitaryInfo.nID = pRecvPacket->strtMilitaryInfo.nID;
    pSendPacket->strtMilitaryInfo.nTotalArtillery = pRecvPacket->strtMilitaryInfo.nTotalArtillery;
    pSendPacket->strtMilitaryInfo.nUnitHealth = pRecvPacket->strtMilitaryInfo.nUnitHealth;
    pSendPacket->strtMilitaryInfo.nXpos = pRecvPacket->strtMilitaryInfo.nXpos;
    pSendPacket->strtMilitaryInfo.nYpos = pRecvPacket->strtMilitaryInfo.nYpos;
    pSendPacket->nIsFin = pRecvPacket->nIsFin;

}

// ********************************************************************
// * 함 수 명: CreateRequestOrderInSallyPacket
// * 설    명: 부대에 자주포 출격명령 요청을 한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 08. 01
// ********************************************************************
void PacketUtil::CreateRequestOrderInSallyPacket(int32_t nThisFd, int32_t nClientFd, int16_t nMilitaryId, int32_t nDstX, int32_t nDstY, StRequestOrderInSallyPacket* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::REQUEST_ORDER_IN_SALLY;
    pStrtPacket->strtHeader.nID = htonl(nThisFd);
    pStrtPacket->nUserClientFd = htonl(nClientFd);
    pStrtPacket->nUserMilitarySection = htons(nMilitaryId);
    pStrtPacket->nDstXPos = htonl(nDstX);
    pStrtPacket->nDstYPos = htonl(nDstY);
}

// ********************************************************************
// * 함 수 명: CreateResponseOrderInSallyPacket
// * 설    명: 부대에 자주포 출격명령 요청에 대한 응답을 한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 08. 01
// ********************************************************************
void PacketUtil::CreateResponseOrderInSallyPacket(int32_t nClientFd, const StResponseOrderInSallyPacket* pRecvPacket, StResponseOrderInSallyPacket* pSendPacket)
{
    pSendPacket->strtHeader.nID = htonl(nClientFd);
    pSendPacket->strtHeader.nType = PACKET_TYPE::RESPONSE_ORDER_IN_SALLY;
    pSendPacket->nUserClientFd = pRecvPacket->nUserClientFd;

    //부대정보
    pSendPacket->strtMilitaryInfo.nID = pRecvPacket->strtMilitaryInfo.nID;
    pSendPacket->strtMilitaryInfo.nTotalArtillery = pRecvPacket->strtMilitaryInfo.nTotalArtillery;
    pSendPacket->strtMilitaryInfo.nUnitHealth = pRecvPacket->strtMilitaryInfo.nUnitHealth;
    pSendPacket->strtMilitaryInfo.nXpos = pRecvPacket->strtMilitaryInfo.nXpos;
    pSendPacket->strtMilitaryInfo.nYpos = pRecvPacket->strtMilitaryInfo.nYpos;

    //자주포정보
    pSendPacket->strtArtilleryInfo.nID = pRecvPacket->strtArtilleryInfo.nID;
    pSendPacket->strtArtilleryInfo.nTotalMissile = htonl(pRecvPacket->strtArtilleryInfo.nTotalMissile);
    pSendPacket->strtArtilleryInfo.nUnitHealth = pRecvPacket->strtArtilleryInfo.nUnitHealth;
    pSendPacket->strtArtilleryInfo.nXpos = pRecvPacket->strtArtilleryInfo.nXpos;
    pSendPacket->strtArtilleryInfo.nYpos = pRecvPacket->strtArtilleryInfo.nYpos;
}

// ********************************************************************
// * 함 수 명: CreateRequestOrderInAttackPacket
// * 설    명: 부대에 missile attack명령 요청 packet 정의
// * 작 성 자: KJH
// * 작성날짜: 2025. 09. 08
// ********************************************************************
void PacketUtil::CreateRequestOrderInAttackPacket(int32_t nThisFd, int32_t nClientFd, int16_t nArtilleryId, int32_t nDstX, int32_t nDstY, StRequestOrderInAttackPacket* pStrtPacket)
{
    pStrtPacket->strtHeader.nType = PACKET_TYPE::REQUEST_ORDER_IN_ATTACK;
    pStrtPacket->strtHeader.nID = htonl(nThisFd);
    pStrtPacket->nUserClientFd = htonl(nClientFd);
    pStrtPacket->nArtilleryId = htons(nArtilleryId);
    pStrtPacket->nDstXPos = htonl(nDstX);
    pStrtPacket->nDstYPos = htonl(nDstY);

}

// ********************************************************************
// * 함 수 명: CreateResponseOrderInAttackPacket
// * 설    명: 부대에 missile attack명령 요청에 대한 응답 패킷 정의
// * 작 성 자: KJH
// * 작성날짜: 2025. 09. 08
// ********************************************************************
void PacketUtil::CreateResponseOrderInAttackPacket(int32_t nClientFd, const StResponseOrderInAttackPacket* pRecvPacket, StResponseOrderInAttackPacket* pSendPacket)
{
}

// ********************************************************************
// * 함 수 명: CreateSyncArtilleryCoordinatePacket
// * 설    명: 자주포위치를 시뮬레이션 엔진과 동기화한다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 08. 11
// ********************************************************************
void PacketUtil::CreateSyncArtilleryCoordinatePacket(int32_t nClientFd, const StSyncArtilleryCoordinatePacket* pRecvPacket, StSyncArtilleryCoordinatePacket* pSendPacket)
{
    pSendPacket->strtHeader.nID = htonl(nClientFd);
    pSendPacket->strtHeader.nType = PACKET_TYPE::SYNC_ARTILLERY_COORDINATE;
    pSendPacket->strtHeader.nFlagField = pRecvPacket->strtHeader.nFlagField;
    pSendPacket->nUserClientFd = pRecvPacket->nUserClientFd;

    //자주포정보
    pSendPacket->strtArtilleryInfo.nID = pRecvPacket->strtArtilleryInfo.nID;
    pSendPacket->strtArtilleryInfo.nTotalMissile = htonl(pRecvPacket->strtArtilleryInfo.nTotalMissile);
    pSendPacket->strtArtilleryInfo.nUnitHealth = pRecvPacket->strtArtilleryInfo.nUnitHealth;
    pSendPacket->strtArtilleryInfo.nXpos = pRecvPacket->strtArtilleryInfo.nXpos;
    pSendPacket->strtArtilleryInfo.nYpos = pRecvPacket->strtArtilleryInfo.nYpos;
}
// ********************************************************************
// * 함 수 명: CreateSyncMissileCoordinatePacket
// * 설    명: missile객체 위치값 동기화 패킷 정의
// * 작 성 자: KJH
// * 작성날짜: 2025. 09. 08
// ********************************************************************
void PacketUtil::CreateSyncMissileCoordinatePacket(int32_t nClientFd, const StSyncMissileCoordinatePacket* pRecvPacket, StSyncMissileCoordinatePacket* pSendPacket)
{
    pSendPacket->strtHeader.nID = htonl(nClientFd);
    pSendPacket->strtHeader.nType = PACKET_TYPE::SYNC_MISSILE_COORDINATE;
    pSendPacket->strtHeader.nFlagField = pRecvPacket->strtHeader.nFlagField;
    pSendPacket->nUserClientFd = pRecvPacket->nUserClientFd;

    //missile정보
    pSendPacket->strtMissileInfo.nID = pRecvPacket->strtMissileInfo.nID;
    pSendPacket->strtMissileInfo.nXpos = pRecvPacket->strtMissileInfo.nXpos;
    pSendPacket->strtMissileInfo.nYpos = pRecvPacket->strtMissileInfo.nYpos;

}

// ********************************************************************
// * 함 수 명: CreateSyncSimulationTimePacket
// * 설    명: broadcast 로 모든 클라이언트에게 현재 시뮬레이션 시각을 알린다.
// * 작 성 자: KJH
// * 작성날짜: 2025. 08. 11
// ********************************************************************
void PacketUtil::CreateSyncSimulationTimePacket(StSyncSimulationTimePacket* pSendPacket, const StSyncSimulationTimePacket* pRecvPacket)
{
    pSendPacket->strtHeader.nType = PACKET_TYPE::SYNC_SIM_TIME;
    pSendPacket->nCurrentSimulationTime = pRecvPacket->nCurrentSimulationTime;
}